src/table/use-sticky-columns.tsx

```diff
@@ -0,0 +1,224 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useLayoutEffect, useState, createRef, useEffect, useCallback, useMemo } from 'react';
+import { TableProps } from './interfaces';
+import { CellOffsets } from './internal';
+import clsx from 'clsx';
+interface StickyStyles {
+  sticky: {
+    left?: string;
+    right?: string;
+  };
+  stuck: {
+    paddingLeft?: string;
+  };
+}
+
+interface StickyColumnParams {
+  containerWidth: number | null;
+  hasSelection: boolean;
+  isWrapperScrollable: boolean;
+  stickyColumns?: TableProps.StickyColumns;
+  visibleColumnsLength: number;
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}
+
+export interface GetStickyColumnProperties {
+  isSticky: boolean;
+  isLastStickyLeft: boolean;
+  isLastStickyRight: boolean;
+  stickyStyles: StickyStyles;
+}
+
+// We allow the table to have a minimum of 148px of available space besides the sum of the widths of the sticky columns
+// This value is an UX recommendation and is approximately 1/3 of our smallest breakpoint (465px)
+const MINIMUM_SCROLLABLE_SPACE = 148;
+
+const useCellOffsets = (tableCellRefs: Array<React.RefObject<HTMLTableCellElement>>) => {
+  // This hook calculates the cumulative offsets of the cells in each column of the table, based on the widths of their siblings
+
+  const [cellOffsets, setCellOffsets] = useState<CellOffsets>({ first: [], last: [] });
+
+  const updateCellOffsets = useCallback(() => {
+    // Calculate widths of all previous siblings of each table cell in the `tableCellRefs` array
+    let firstColumnsOffsets = tableCellRefs
+      .map(ref => (ref?.current?.previousSibling as HTMLTableCellElement)?.offsetWidth)
+      .filter(Boolean);
+    // Calculate cumulative widths of previous siblings to get the total offset of each of the cells
+    firstColumnsOffsets = firstColumnsOffsets.map((elem, index) =>
+      firstColumnsOffsets.slice(0, index + 1).reduce((a, b) => a + b)
+    );
+
+    // Calculate widths of all next siblings of each table cell in the `tableCellRefs` array
+    let lastColumnsOffsets = tableCellRefs.map(ref => (ref?.current?.nextSibling as HTMLTableCellElement)?.offsetWidth);
+    lastColumnsOffsets = lastColumnsOffsets.filter(Boolean).reverse();
+    // Calculate cumulative widths of next siblings to get the the total offset of each of the cells
+    lastColumnsOffsets = lastColumnsOffsets
+      .map((elem, index) => lastColumnsOffsets.slice(0, index + 1).reduce((a, b) => a + b))
+      .reverse();
+
+    setCellOffsets({ first: [0, ...firstColumnsOffsets], last: [...lastColumnsOffsets, 0] });
+  }, [tableCellRefs]);
+
+  useLayoutEffect(() => {
+    // Request animation frame to make sure effect runs after the browser's automatic table layout algorithm
+    requestAnimationFrame(() => {
```

Comment:
Should we cancel the animation frame in the effect's cleanup? Otherwise, the updates can stack or even fire after the component is un-mounted.